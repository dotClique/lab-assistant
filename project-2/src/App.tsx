import React from 'react';
import Header from "./components/Header";
import Tab from './components/Tab';
import Info from './components/Info';

function App() {
  return (
    <div>
      <Header />
      <Info />
      <Tab />
    </div>
  );
}

export default App;
