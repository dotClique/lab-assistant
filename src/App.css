.authFormContainer {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    margin: 2em;
}

.authForm {
    max-width: 500px;
}