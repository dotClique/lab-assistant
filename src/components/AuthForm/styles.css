.auth-form {
    width: 100%;
}

.auth-form-info {
    margin-bottom: 2em;
}