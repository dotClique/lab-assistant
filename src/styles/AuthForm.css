.authForm {
    width: 100%;
}

.authFormInfo {
    margin-bottom: 2em;
}