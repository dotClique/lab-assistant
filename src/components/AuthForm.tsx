import React from "react";
import {Alert, Button, Form, Input, Typography} from "antd";
import {AuthContext, AuthContextType} from "../App";
import {isAuthorized} from "../api/ApiBase";
import '../styles/AuthForm.css';

const {Paragraph, Text} = Typography;

// Define the types of the content of a TokenInputState, aka. form submission content
interface TokenInputState {
    inputAccessToken: string,
    inputProjectId: string,
    authFailed: boolean
}

// Class component that renders the authorization form in the application
export default class AuthForm extends React.Component<{}, TokenInputState> {

    constructor(props: {}) {
        super(props);
        this.state = {
            inputAccessToken: "",
            inputProjectId: "",
            authFailed: false
        }
        this.submitInput = this.submitInput.bind(this);
        this.onFinishFailed = this.onFinishFailed.bind(this);
    }

    // Submit the input from the authorization form the user is first prompted with in the app
    // Based on the authorization-method in ApiBase.ts
    async submitInput(auth: AuthContextType, input: {projectId: string, accessToken: string}) {
        const projectId = input.projectId
        const accessToken = input.accessToken
        if (await isAuthorized(accessToken, projectId)) {
            this.setState({authFailed: false})
            auth.setAccessToken(accessToken)
            auth.setProjectId(projectId)
            auth.setAuthorized(true)
        } else {
            this.setState({authFailed: true})
        }
    }

    // If the form fails on submission, log the error in the console
    onFinishFailed(error: any) {
        console.log('Failed:', error);
    };

    render() {
        return (
            <AuthContext.Consumer>
                {auth => (
                    <>
                        <Alert className={"auth-form-info"} type="info" message="Welcome!"
                               description={
                                   <Text>
                                       <Paragraph>
                                           In order to display some cool insights into your project, we need valid
                                           credentials for accessing your GitLab project.
                                       </Paragraph>
                                       <Paragraph>
                                           Your project identification can either be the number displayed right under your project name on GitLab,
                                           or the project path visible in the URL (the part after <Text code>https://gitlab.stud.idi.ntnu.no/</Text>)
                                       </Paragraph>
                                       <Paragraph>
                                           An access token can be generated from GitLab at <Text code>Settings &gt; Access Tokens</Text>
                                       </Paragraph>
                                   </Text>
                               }
                               showIcon/>
                        <Form
                            onFinish={(input) => this.submitInput(auth, input)}
                            onFinishFailed={this.onFinishFailed}
                            layout={"vertical"}
                        >
                            <Form.Item
                                label="Project Identification"
                                name="projectId"
                                rules={[{required: true, message: 'Please input the project identification!'}]}
                            >
                                <Input placeholder={"e.g. '12345' or 'it2810-h21/team-321/project-2'"}/>
                            </Form.Item>

                            <Form.Item
                                label="Access Token"
                                name="accessToken"
                                rules={[{required: true, message: 'Please input your access token!'}]}
                            >
                                <Input.Password placeholder={"e.g. 'f3bba1Qw7MjrbvwxLGUn'"}/>
                            </Form.Item>

                            <Form.Item>
                                <Button type="primary" htmlType="submit">
                                    Submit
                                </Button>
                            </Form.Item>
                        </Form>
                        {this.state.authFailed && (
                            <Alert
                                message="Authentication failed"
                                description="The project does not exist, or you are not authorized to access it :("
                                type="warning"
                            />
                        )
                        }
                    </>
                )
                }
            </AuthContext.Consumer>
        )
    }
}
