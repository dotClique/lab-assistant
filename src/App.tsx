import React, {useEffect, useState} from 'react';
import Header from "./components/Header";
import Tab from './components/Tab';
import Info from './components/Info';
import Page from "./components/Page";
import {getNames} from "./api/Users";
import {getEmoji} from "./api/Awards";
import './App.css';
import AuthForm from "./components/AuthForm";
import {
    getLocalTheme,
    setLocalTheme,
    clearSessionCredentials,
    getSessionAccessToken,
    getSessionProjectId,
    sessionHasCredentials,
    setSessionAccessToken,
    setSessionProjectId,
    getLocalInfoViewedStatus
} from "./webstorage/WebStorage";
import {isAuthorized} from "./api/ApiBase";

// Define the auth context of the application
export const AuthContext = React.createContext({
    authorized: false,
    setAuthorized: (authorized: boolean) => {
    },
    accessToken: "",
    setAccessToken: (accessToken: string) => {
    },
    projectId: "",
    setProjectId: (projectId: string) => {
    }
})

// Define the types of the auth context
export interface AuthContextType {
    authorized: boolean,
    setAuthorized: (authorized: boolean) => void,
    accessToken: string,
    setAccessToken: (accessToken: string) => void,
    projectId: string,
    setProjectId: (projectId: string) => void
}

// Define the theme context of the application
export const ThemeContext = React.createContext<{ theme: string, toggleTheme: () => void }>({
    theme: "orange",
    toggleTheme: () => {
    }
})

// Define the types of the assets context
export interface AssetsContextType {
    names: {
        nouns: string[],
        adjectives: string[],
    },
    emoji: {
        [name: string]: string,
    }
}
// Define the assets context of the application
export const AssetsContext = React.createContext<AssetsContextType>({
    names: {nouns: ["unknown"], adjectives: [""]},
    emoji: {}
});

// Main function
function App() {
    const localTheme = getLocalTheme();

    const sessionProjectId = getSessionProjectId();
    const sessionAccessToken = getSessionAccessToken();

    if (sessionAccessToken === "" || sessionProjectId === "") {
        // Empty access token and project id will be accepted in requests, so deny it here
        clearSessionCredentials()
    }

    const [{authorized, accessToken, projectId, theme, assets}, setState] = useState<{
        authorized: boolean,
        accessToken: string,
        projectId: string,
        theme: string,
        assets: AssetsContextType,
    }>({
        authorized: sessionHasCredentials(),
        accessToken: sessionAccessToken != null ? sessionAccessToken : "",
        projectId: sessionProjectId != null ? sessionProjectId : "",
        theme: localTheme ?? "orange",
        assets: {
            names: {
                nouns: ["unknown"], adjectives: [""]
            },
            emoji: {},
        }
    });

    /**
    * Toggles the website theme to be opposite of what it last was
    */
    const toggleTheme = () => {
        const newTheme = theme === "orange" ? "blue" : "orange";
        setState(prev => ({...prev, theme: newTheme}))
        setLocalTheme(newTheme);
    }

    /**
     * Close info welcome message permanently after first close
     * @returns boolean
     */
    const closeInfo = () => {
        if (getLocalInfoViewedStatus() === "true") {
            return false;
        }
        else return true;
    }

    useEffect(() => {
        getNames().then(names => setState(prev => ({...prev, assets: {...prev.assets, names: names}})))
        getEmoji().then(emoji => setState(prev => ({...prev, assets: {...prev.assets, emoji: emoji}})))
    }, []);

    /**
     * Verify API credentials, and redirect to auth form in case they are invalid
     */
    useEffect(() => {
        if (accessToken === "" || projectId === "") {
            // Credentials not set, no need to verify them
            return
        }
        isAuthorized(accessToken, projectId).then(authorized => {
            setState(prev => ({...prev, authorized: authorized}));
            if (!authorized) {
                clearSessionCredentials()
            }
        })
    }, [accessToken, projectId])

    const authContextProviderValue: AuthContextType = {
        authorized: authorized,
        setAuthorized: authorized => {
            setState(prevState => ({...prevState, authorized: authorized}));
            if (!authorized) {
                clearSessionCredentials()
            }
        },
        accessToken: accessToken,
        setAccessToken: accessToken => {
            setState(prevState => ({...prevState, accessToken: accessToken}))
            setSessionAccessToken(accessToken)
        },
        projectId: projectId,
        setProjectId: projectId => {
            setState(prevState => ({...prevState, projectId: projectId}))
            setSessionProjectId(projectId)
        },
    }

    return (
        <AuthContext.Provider value={authContextProviderValue}>
            <ThemeContext.Provider value={{theme: theme, toggleTheme: toggleTheme}}>
                <Page>
                    <Header/>
                    {!authorized ?
                        (
                            <div className={"authFormContainer"}>
                                <div className={"auth-form"}>
                                    <AuthForm/>
                                </div>
                            </div>
                        ) : (
                            <AssetsContext.Provider value={assets}>
                                {closeInfo() &&
                                    <Info/>}
                                <Tab/>
                            </AssetsContext.Provider>
                        )
                    }
                </Page>
            </ThemeContext.Provider>
        </AuthContext.Provider>
    );
}

export default App;
