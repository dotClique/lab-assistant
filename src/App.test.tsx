import React from 'react';
import App from './App';
import {render, screen, waitFor, within} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import AuthForm from "./components/AuthForm";

// Fix for "TypeError: window.matchMedia is not a function"
// ref: https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
window.matchMedia = (query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
})

it("matches snapshot", () => {
    expect(render(<App/>)).toMatchSnapshot();
});

it('uses orange theme by default', () => {
    render(<App/>)
    const header = screen.getByRole('banner')
    expect(header).toHaveClass('orange')
    expect(header).not.toHaveClass('blue')
})

it('changes theme on theme switch toggle', () => {
    render(<App/>)
    const header = screen.getByRole('banner')
    userEvent.click(within(header).getByRole('switch'))
    expect(header).toHaveClass('blue')
    expect(header).not.toHaveClass('orange')
    userEvent.click(within(header).getByRole('switch'))
    expect(header).toHaveClass('orange')
    expect(header).not.toHaveClass('blue')
})

it('allows typing in auth form', () => {
    render(<AuthForm/>)
    const projectId = "thisissomeprojectid"
    const projectIdInput = screen.getByLabelText("projectIdInput")
    userEvent.type(projectIdInput, projectId)
    expect(projectIdInput).toHaveValue(projectId)
    const accessToken = "thisissurelyavalidaccesstoken"
    const accessTokenInput = screen.getByLabelText("accessTokenInput")
    userEvent.type(accessTokenInput, accessToken)
    expect(accessTokenInput).toHaveValue(accessToken)
})