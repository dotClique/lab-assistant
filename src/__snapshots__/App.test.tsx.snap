// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="page"
      >
        <div
          class="header orange"
          role="banner"
        >
          <img
            alt="logo"
            class="App-logo"
            src="gitlab-logo-w.png"
          />
          <div
            class="header-text-container"
          >
            <h1
              class="header-text"
            >
              Lab Assistant
            </h1>
          </div>
          <div
            class="theme-switch-container"
          >
            <span
              aria-label="bg-colors"
              class="anticon anticon-bg-colors theme-switch-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="bg-colors"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M766.4 744.3c43.7 0 79.4-36.2 79.4-80.5 0-53.5-79.4-140.8-79.4-140.8S687 610.3 687 663.8c0 44.3 35.7 80.5 79.4 80.5zm-377.1-44.1c7.1 7.1 18.6 7.1 25.6 0l256.1-256c7.1-7.1 7.1-18.6 0-25.6l-256-256c-.6-.6-1.3-1.2-2-1.7l-78.2-78.2a9.11 9.11 0 00-12.8 0l-48 48a9.11 9.11 0 000 12.8l67.2 67.2-207.8 207.9c-7.1 7.1-7.1 18.6 0 25.6l255.9 256zm12.9-448.6l178.9 178.9H223.4l178.8-178.9zM904 816H120c-4.4 0-8 3.6-8 8v80c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-80c0-4.4-3.6-8-8-8z"
                />
              </svg>
            </span>
            <button
              aria-checked="false"
              class="ant-switch theme-switch orange"
              role="switch"
              type="button"
            >
              <div
                class="ant-switch-handle"
              />
              <span
                class="ant-switch-inner"
              />
            </button>
          </div>
        </div>
        <div
          class="authFormContainer"
        >
          <div
            class="auth-form"
          >
            <div
              class="ant-alert ant-alert-info ant-alert-with-description auth-form-info"
              data-show="true"
              role="alert"
            >
              <span
                aria-label="info-circle"
                class="anticon anticon-info-circle ant-alert-icon"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="info-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  />
                  <path
                    d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
                  />
                </svg>
              </span>
              <div
                class="ant-alert-content"
              >
                <div
                  class="ant-alert-message"
                >
                  Welcome!
                </div>
                <div
                  class="ant-alert-description"
                >
                  <span
                    class="ant-typography"
                  >
                    <div
                      class="ant-typography"
                    >
                      In order to display some cool insights into your project, we need valid credentials for accessing your GitLab project.
                    </div>
                    <div
                      class="ant-typography"
                    >
                      Your project identification can either be the number displayed right under your project name on GitLab, or the project path visible in the URL (the part after 
                      <span
                        class="ant-typography"
                      >
                        <code>
                          https://gitlab.stud.idi.ntnu.no/
                        </code>
                      </span>
                      )
                    </div>
                    <div
                      class="ant-typography"
                    >
                      An access token can be generated from GitLab at 
                      <span
                        class="ant-typography"
                      >
                        <code>
                          Settings &gt; Access Tokens
                        </code>
                      </span>
                    </div>
                  </span>
                </div>
              </div>
            </div>
            <form
              class="ant-form ant-form-vertical"
            >
              <div
                class="ant-row ant-form-item"
              >
                <div
                  class="ant-col ant-form-item-label"
                >
                  <label
                    class="ant-form-item-required"
                    for="projectId"
                    title="Project Identification"
                  >
                    Project Identification
                  </label>
                </div>
                <div
                  class="ant-col ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <input
                        aria-label="projectIdInput"
                        class="ant-input"
                        id="projectId"
                        placeholder="e.g. '12345' or 'it2810-h21/team-321/project-2'"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="ant-row ant-form-item"
              >
                <div
                  class="ant-col ant-form-item-label"
                >
                  <label
                    class="ant-form-item-required"
                    for="accessToken"
                    title="Access Token"
                  >
                    Access Token
                  </label>
                </div>
                <div
                  class="ant-col ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <span
                        class="ant-input-affix-wrapper ant-input-password"
                      >
                        <input
                          action="click"
                          aria-label="accessTokenInput"
                          class="ant-input"
                          id="accessToken"
                          placeholder="e.g. 'f3bba1Qw7MjrbvwxLGUn'"
                          type="password"
                          value=""
                        />
                        <span
                          class="ant-input-suffix"
                        >
                          <span
                            aria-label="eye-invisible"
                            class="anticon anticon-eye-invisible ant-input-password-icon"
                            role="img"
                            tabindex="-1"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="eye-invisible"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M942.2 486.2Q889.47 375.11 816.7 305l-50.88 50.88C807.31 395.53 843.45 447.4 874.7 512 791.5 684.2 673.4 766 512 766q-72.67 0-133.87-22.38L323 798.75Q408 838 512 838q288.3 0 430.2-300.3a60.29 60.29 0 000-51.5zm-63.57-320.64L836 122.88a8 8 0 00-11.32 0L715.31 232.2Q624.86 186 512 186q-288.3 0-430.2 300.3a60.3 60.3 0 000 51.5q56.69 119.4 136.5 191.41L112.48 835a8 8 0 000 11.31L155.17 889a8 8 0 0011.31 0l712.15-712.12a8 8 0 000-11.32zM149.3 512C232.6 339.8 350.7 258 512 258c54.54 0 104.13 9.36 149.12 28.39l-70.3 70.3a176 176 0 00-238.13 238.13l-83.42 83.42C223.1 637.49 183.3 582.28 149.3 512zm246.7 0a112.11 112.11 0 01146.2-106.69L401.31 546.2A112 112 0 01396 512z"
                              />
                              <path
                                d="M508 624c-3.46 0-6.87-.16-10.25-.47l-52.82 52.82a176.09 176.09 0 00227.42-227.42l-52.82 52.82c.31 3.38.47 6.79.47 10.25a111.94 111.94 0 01-112 112z"
                              />
                            </svg>
                          </span>
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="ant-row ant-form-item"
              >
                <div
                  class="ant-col ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <button
                        aria-label="submitCredentials"
                        class="ant-btn ant-btn-primary"
                        type="submit"
                      >
                        <span>
                          Submit
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="page"
    >
      <div
        class="header orange"
        role="banner"
      >
        <img
          alt="logo"
          class="App-logo"
          src="gitlab-logo-w.png"
        />
        <div
          class="header-text-container"
        >
          <h1
            class="header-text"
          >
            Lab Assistant
          </h1>
        </div>
        <div
          class="theme-switch-container"
        >
          <span
            aria-label="bg-colors"
            class="anticon anticon-bg-colors theme-switch-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="bg-colors"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M766.4 744.3c43.7 0 79.4-36.2 79.4-80.5 0-53.5-79.4-140.8-79.4-140.8S687 610.3 687 663.8c0 44.3 35.7 80.5 79.4 80.5zm-377.1-44.1c7.1 7.1 18.6 7.1 25.6 0l256.1-256c7.1-7.1 7.1-18.6 0-25.6l-256-256c-.6-.6-1.3-1.2-2-1.7l-78.2-78.2a9.11 9.11 0 00-12.8 0l-48 48a9.11 9.11 0 000 12.8l67.2 67.2-207.8 207.9c-7.1 7.1-7.1 18.6 0 25.6l255.9 256zm12.9-448.6l178.9 178.9H223.4l178.8-178.9zM904 816H120c-4.4 0-8 3.6-8 8v80c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-80c0-4.4-3.6-8-8-8z"
              />
            </svg>
          </span>
          <button
            aria-checked="false"
            class="ant-switch theme-switch orange"
            role="switch"
            type="button"
          >
            <div
              class="ant-switch-handle"
            />
            <span
              class="ant-switch-inner"
            />
          </button>
        </div>
      </div>
      <div
        class="authFormContainer"
      >
        <div
          class="auth-form"
        >
          <div
            class="ant-alert ant-alert-info ant-alert-with-description auth-form-info"
            data-show="true"
            role="alert"
          >
            <span
              aria-label="info-circle"
              class="anticon anticon-info-circle ant-alert-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="info-circle"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                />
                <path
                  d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
                />
              </svg>
            </span>
            <div
              class="ant-alert-content"
            >
              <div
                class="ant-alert-message"
              >
                Welcome!
              </div>
              <div
                class="ant-alert-description"
              >
                <span
                  class="ant-typography"
                >
                  <div
                    class="ant-typography"
                  >
                    In order to display some cool insights into your project, we need valid credentials for accessing your GitLab project.
                  </div>
                  <div
                    class="ant-typography"
                  >
                    Your project identification can either be the number displayed right under your project name on GitLab, or the project path visible in the URL (the part after 
                    <span
                      class="ant-typography"
                    >
                      <code>
                        https://gitlab.stud.idi.ntnu.no/
                      </code>
                    </span>
                    )
                  </div>
                  <div
                    class="ant-typography"
                  >
                    An access token can be generated from GitLab at 
                    <span
                      class="ant-typography"
                    >
                      <code>
                        Settings &gt; Access Tokens
                      </code>
                    </span>
                  </div>
                </span>
              </div>
            </div>
          </div>
          <form
            class="ant-form ant-form-vertical"
          >
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-form-item-label"
              >
                <label
                  class="ant-form-item-required"
                  for="projectId"
                  title="Project Identification"
                >
                  Project Identification
                </label>
              </div>
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <input
                      aria-label="projectIdInput"
                      class="ant-input"
                      id="projectId"
                      placeholder="e.g. '12345' or 'it2810-h21/team-321/project-2'"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-form-item-label"
              >
                <label
                  class="ant-form-item-required"
                  for="accessToken"
                  title="Access Token"
                >
                  Access Token
                </label>
              </div>
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <span
                      class="ant-input-affix-wrapper ant-input-password"
                    >
                      <input
                        action="click"
                        aria-label="accessTokenInput"
                        class="ant-input"
                        id="accessToken"
                        placeholder="e.g. 'f3bba1Qw7MjrbvwxLGUn'"
                        type="password"
                        value=""
                      />
                      <span
                        class="ant-input-suffix"
                      >
                        <span
                          aria-label="eye-invisible"
                          class="anticon anticon-eye-invisible ant-input-password-icon"
                          role="img"
                          tabindex="-1"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="eye-invisible"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M942.2 486.2Q889.47 375.11 816.7 305l-50.88 50.88C807.31 395.53 843.45 447.4 874.7 512 791.5 684.2 673.4 766 512 766q-72.67 0-133.87-22.38L323 798.75Q408 838 512 838q288.3 0 430.2-300.3a60.29 60.29 0 000-51.5zm-63.57-320.64L836 122.88a8 8 0 00-11.32 0L715.31 232.2Q624.86 186 512 186q-288.3 0-430.2 300.3a60.3 60.3 0 000 51.5q56.69 119.4 136.5 191.41L112.48 835a8 8 0 000 11.31L155.17 889a8 8 0 0011.31 0l712.15-712.12a8 8 0 000-11.32zM149.3 512C232.6 339.8 350.7 258 512 258c54.54 0 104.13 9.36 149.12 28.39l-70.3 70.3a176 176 0 00-238.13 238.13l-83.42 83.42C223.1 637.49 183.3 582.28 149.3 512zm246.7 0a112.11 112.11 0 01146.2-106.69L401.31 546.2A112 112 0 01396 512z"
                            />
                            <path
                              d="M508 624c-3.46 0-6.87-.16-10.25-.47l-52.82 52.82a176.09 176.09 0 00227.42-227.42l-52.82 52.82c.31 3.38.47 6.79.47 10.25a111.94 111.94 0 01-112 112z"
                            />
                          </svg>
                        </span>
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <button
                      aria-label="submitCredentials"
                      class="ant-btn ant-btn-primary"
                      type="submit"
                    >
                      <span>
                        Submit
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
